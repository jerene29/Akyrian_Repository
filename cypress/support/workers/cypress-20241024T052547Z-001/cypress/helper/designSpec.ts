import { GetStudyRevisionListDocument } from '../../src/graphQL/generated/graphql';
import { mockUserDataAdmin } from '../../src/constant/testFixtures';

type SpecNumber = `S-${number}-${number}`;

type SpecInfo = {
  // e.g S-09-010
  spec: SpecNumber;
  // title e.g headerSection, combined description becomes S-09-010 headerSection, but optional
  title?: string;
};

type Options = {
  loginAs?: string;
  skipTest?: boolean;
};

function resolveDescription(description: SpecNumber | SpecInfo) {
  return typeof description === 'string' ? description : `${description.spec} ${description.title}`;
}

const aliasGetStudyRevisionListDocument =
  'name' in GetStudyRevisionListDocument.definitions[0]
    ? GetStudyRevisionListDocument.definitions[0].name?.value
    : 'GetStudyRevisionListDocument';
/*
example usage
designSpec('S-19-2', () => {

  describe('...', ...)

  it('checks something...', () => {...})
});
*/
export function designSpec(
  description: SpecNumber | SpecInfo,
  callback: () => void,
  options?: Options,
) {
  before(() => {
    cy.intercept('POST', '/graphql', (req) => {
      if (req.body.operationName === aliasGetStudyRevisionListDocument) {
        req.alias = req.body.operationName;
      }
    });
    cy.beforeSetup(mockUserDataAdmin);
    cy.clearLocalStorageSnapshot();
    cy.saveLocalStorage();
    cy.wait(`@${aliasGetStudyRevisionListDocument}`);
  });
  if (options?.skipTest) {
    describe.skip(resolveDescription(description), callback);
  } else if (typeof description !== 'string' && description.title) {
    describe(description.spec, () => {
      describe(String(description.title), callback);
    });
  } else {
    describe(resolveDescription(description), callback);
  }
}
