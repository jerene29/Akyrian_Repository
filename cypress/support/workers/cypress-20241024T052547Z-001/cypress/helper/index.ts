export * from './designSpec';
export * from './object';