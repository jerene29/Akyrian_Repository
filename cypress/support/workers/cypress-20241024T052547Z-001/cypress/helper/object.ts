/**
 * to shorten the tedious '[data-cy=selector]' selector.
 * just put the name of data-cy attributed element.
 * -- d stands for document
 *
```
d`addButton`
```
 */
export function d(args: TemplateStringsArray) {
  return `[data-cy=${args.join('')}]`;
}
