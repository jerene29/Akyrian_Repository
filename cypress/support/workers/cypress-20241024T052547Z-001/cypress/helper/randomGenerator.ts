import moment = require("moment");

export const randomAlphaNumeric = (length: number) => {
  const chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
  let result = '';
  for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
  return result;
}

export const randomAlphabet = (length: number) => {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
  let result = '';
  for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
  return result;
}

export const randomDate = () => {
  const start = new Date(2012, 0, 1)
  const end = new Date()
  const date = new Date(start.getTime() + Math.random() * (end.getTime() - start.getTime()))
  return moment(date);
}