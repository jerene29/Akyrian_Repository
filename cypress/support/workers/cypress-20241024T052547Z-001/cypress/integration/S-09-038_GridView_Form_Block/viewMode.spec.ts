import { d, designSpec } from '../../helper';

const options = {
  loginAs: 'admin@example.com',
  skipTest: true,
};

designSpec(
  'S-09-038',
  () => {
    before(() => {
      cy.createStudy({
        name: 'GRIDVIEW_TEST',
        // remember the initial form was 5
        formCount: 5,
        visitCount: 5,
      });
    });

    context('inactive edit form mode', () => {
      const THIRD_FORM = '3-0';

      it('is hoverable', () => {
        // NOTE: data-cy prop for formHeader gone missing on the next day!!! idk why, temporarily use id (#)
        cy.hover('#formHeader-1-0');
        cy.shouldBeVisible(d`editableFormName-1-0`);
      });

      it('reveals text edit details to redirect to edit view', () => {
        cy.hover('#formHeader-1-0');
        cy.shouldBeVisible(d`formEditDetails-1-0`).click();
        // snapshot detail view button on detail view, to make sure we're there (since it's based on state and not routing).
        cy.get(d`detail-view`).should('have.css', 'background-color', 'rgb(61, 97, 215)');
        cy.get(d`grid-view`).click();
      });

      it('reveals delete icon and can delete form', () => {
        cy.shouldBeVisible(d`formDelete-1-0`).click();
        cy.shouldBeVisible(d`confirmation-modal-desc`);

        cy.shouldBeVisible(d`confirmModal-confirmButton`).click();
        // check if form on row 5 still there.
        cy.isGone(d`formDelete-5-0`);
      });

      it('reveals + icons to the top and bottom of the block, can add form according to the order of add button position ', () => {
        // NOTE: special quirky operation here because cypress is too fast to get the item id, before the previous deletion is completed
        cy.get(d`detail-view`).click();
        cy.get(d`buttonConfirm`).click();

        cy.hover(`#formHeader-${THIRD_FORM}`);
        cy.shouldBeVisible(`[data-cy=copyFormUpButton-${THIRD_FORM}]`).click();

        // third form became fourth form
        cy.shouldBeVisible(`[data-cy=copyFormDownButton-4-0]`).click();

        cy.get(`#formHeader-3-0`).contains('unnamed form');
      });

      //  NOTE: put to bottommost because to do this we need to test it on seeded gridview data.
      it('checks if delete form remove parent child trigger', () => {
        cy.visit('/study');
        cy.get('#active-study1revisionDev2e').click();
        cy.get(d`btn-edit-study`).click();
        cy.get(d`grid-view`).click();
        cy.hover('#formHeader-3-0');
        cy.shouldBeVisible(d`formDelete-3-0`).click();
        cy.get('.ant-modal-body').contains(
          'You are deleting a form that has dependencies. Deleting this form will disconnect them. Are you sure?',
        );
        cy.shouldBeVisible(d`confirmModal-confirmButton`).click();
      });

      // TODO: rework on this intermittently fail on cloud.
      xit('is draggable', () => {
        cy.dragAndDrop('#formHeader-1-0', '#formHeader-3-0');
        cy.shouldBeVisible(d`confirmModal-confirmButton`).click();
        cy.get('body').tab();
        cy.reload();
        cy.get(d`grid-view`).click();
        cy.get(d`grid-view`).click();
        cy.get(d`grid-view`).click();

        cy.getTextSnapshot('#formHeader-3-0');
      });
    });
  },
  options,
);
