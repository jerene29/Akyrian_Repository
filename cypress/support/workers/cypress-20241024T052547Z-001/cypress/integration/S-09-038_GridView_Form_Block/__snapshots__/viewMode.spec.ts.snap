exports[`S-09-038 > inactive edit form mode > is draggable #0`] =
{
  "snapshot": "Consent Of Subject"
};
